### PR TITLE
AV-1837: General search

### PR DESCRIPTION
**Pull request by teravill**
Implemented a general search that can be used to search for datasets, apisets and showcases. The general search can be accessed via data/search url, from the header's search bar or from the hero on the front page

image
image

### CODE REVIEW

File: ckan/ckanext/ckanext-ytp_main/ckanext/ytp/templates/general_search/snippets/post_pagination.html [hunk 1]
```diff
@@ -0,0 +1,105 @@
+{% asset 'ytp_resources/post_pagination_js' %}
+
+
+{%- set current_page = 1 if not page else page -%}
+{%- set plus_minus = 2 -%}
```
Comment (beazil) [inline ckan/ckanext/ckanext-ytp_main/ckanext/ytp/templates/general_search/snippets/post_pagination.html:5]: This variable isn't quite self descriptive named, so probably rename it or add some comment to describe what is it used for

File: ckan/ckanext/ckanext-ytp_main/ckanext/ytp/views.py [hunk 2]
```diff
@@ -118,6 +124,184 @@ def recent_datasets_feed():
...
+        'views_recent desc'
+    ]
+
+    sort_string = sort_by if sort_by in allowed_sorting else "score desc, metadata_created desc"
+
+    # Organizations require a bit different search term and treatment if they are the only thing searched
+    if dataset_type == 'organization':
+        # if user is searching only for organizations, we want to display more organizations than 5
+        organization_limit = 20
+        dataset_limit = 0
```
Comment (beazil) [inline ckan/ckanext/ckanext-ytp_main/ckanext/ytp/views.py:194]: Can't set this to zero as it is used as a divider later (and it causes an error)

File: drupal/modules/avoindata-header/src/Plugin/Form/SearchForm.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      '#button_type' => 'primary',
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
```
Comment (beazil) [inline drupal/modules/avoindata-header/src/Plugin/Form/SearchForm.php:59]: Is this really needed? If not maybe it should be removed. If it is, maybe describe reason for it with comment

File: ckan/ckanext/ckanext-ytp_main/ckanext/ytp/views.py [hunk 4]
```diff
@@ -118,6 +124,184 @@ def recent_datasets_feed():
...
+    g.general_search = {
+            u'q': q,
+            u'total_results': combined_results,
+            u'result_count': result_count,
+            u'item_count': item_count,
+            u"last_query": params_to_dict(request.form),
+            u'filter': chosen_filter,
+            u'page': page,
+            u'sort_string': sort_string,
+            # u'total_pages': int(math.ceil(float(result_count.get(dataset_type, 0)) / float(dataset_limit + organization_limit))),
```
Comment (beazil) [inline ckan/ckanext/ckanext-ytp_main/ckanext/ytp/views.py:286]: Probably good idea to describe why it's commented (or remove, if there isn't planned usage for this piece of code)

**Review by beazil — CHANGES_REQUESTED** [review]

File: ckan/ckanext/ckanext-ytp_main/ckanext/ytp/views.py (hunk 4 above)
Comment (teravill) [reply ckan/ckanext/ckanext-ytp_main/ckanext/ytp/views.py]: Resolved in [92a479a](https://github.com/vrk-kpa/opendata/pull/1774/commits/92a479a6671ee9d1f0071c475d1135bf7a17fbd2)

File: drupal/modules/avoindata-header/src/Plugin/Form/SearchForm.php (hunk 3 above)
Comment (teravill) [reply drupal/modules/avoindata-header/src/Plugin/Form/SearchForm.php]: Resolved in [92a479a](https://github.com/vrk-kpa/opendata/pull/1774/commits/92a479a6671ee9d1f0071c475d1135bf7a17fbd2)

File: ckan/ckanext/ckanext-ytp_main/ckanext/ytp/views.py (hunk 2 above)
Comment (teravill) [reply ckan/ckanext/ckanext-ytp_main/ckanext/ytp/views.py]: Resolved in [92a479a](https://github.com/vrk-kpa/opendata/pull/1774/commits/92a479a6671ee9d1f0071c475d1135bf7a17fbd2)

File: ckan/ckanext/ckanext-ytp_main/ckanext/ytp/templates/general_search/snippets/post_pagination.html (hunk 1 above)
Comment (teravill) [reply ckan/ckanext/ckanext-ytp_main/ckanext/ytp/templates/general_search/snippets/post_pagination.html]: Resolved in [92a479a](https://github.com/vrk-kpa/opendata/pull/1774/commits/92a479a6671ee9d1f0071c475d1135bf7a17fbd2)

**Review by beazil — APPROVED** [review]

**Pull request merged**